src/riscv_machine.c

```diff
@@ -99,6 +120,190 @@ static uint64_t rtc_get_time(RISCVMachine *m)
     return val;
 }
 
+
+/***************************   UART   ***************************/
+
+#define UART_REG_DATA  0 /* (DLAB=0) RBR=rx buffer / THR=tx holding */
+                         /* (DLAB=1) DLL=divisor latch (lsbyte) */
+#define UART_REG_IER   1 /* (DLAB=0) interrupt enable (bit 0=rx avail, 1=tx empty,
+                                     2=rx line status, 3=modem status) */
+                         /* (DLAB=1) DLM=divisor latch (msbyte) */
+#define UART_REG_IIR   2 /* (read)  interrupt id/cause */
+#define UART_REG_FCR   2 /* (write) FIFO control */
+#define UART_REG_LCR   3 /* line control (bit 7=DLAB) */
+#define UART_REG_MCR   4 /* modem control */
+#define UART_REG_LSR   5 /* line status (bit 0=rx avail, 5=tx buf empty, 6=tx empty/done) */
+#define UART_REG_MSR   6 /* modem status */
+#define UART_REG_SCR   7 /* scratch */
+
+#define UART_LCR_DLAB  0x80
+
+static void uart_set_irq(RISCVMachine *s)
+{
+    int pending = (s->uart_rx_pending && (s->uart_ier & 0x01) != 0)
+                || (s->uart_tx_pending && (s->uart_ier & 0x02) != 0);
+    set_irq(&s->plic_irq[UART_IRQ], pending);
+}
+
+static uint32_t uart_read(void *opaque, uint32_t offset, int size_log2)
+{
+    RISCVMachine *s = (RISCVMachine *)opaque;
+    uint32_t val;
+
+    assert(size_log2 == 0);
+    switch(offset) {
+
+    case UART_REG_DATA:
+        if (s->uart_lcr & UART_LCR_DLAB)
+            val = s->uart_dll;
+        else if (s->uart_rx_head != s->uart_rx_tail)    /* data to read? */
+        {
+            val = s->uart_rx_buf[s->uart_rx_head];
+            s->uart_rx_head = (s->uart_rx_head + 1) % UART_RX_BUFSIZE;
+            if (s->uart_rx_head == s->uart_rx_tail)     /* now empty? */
+            {
+                s->uart_rx_pending = 0;
+                uart_set_irq(s);
+            }
+        }
+        else
+            val = 0;
+        break;
+
+    case UART_REG_IER:
+        val = (s->uart_lcr & UART_LCR_DLAB) ? s->uart_dlm : s->uart_ier;
+        break;
+
+    case UART_REG_LCR:
+        val = s->uart_lcr;
+        break;
+
+    case UART_REG_IIR:
+        val = ((s->uart_fcr & 0x01) ? 0xC0 : 0x00)
+            | (s->uart_rx_pending ? ((s->uart_fcr & 0xC0) ? 0x0C : 0x04) : s->uart_tx_pending ? 0x02 : 0x01);
+        s->uart_rx_pending = 0;
+        s->uart_tx_pending = 0;
+        uart_set_irq(s);
+        break;
+
+    case UART_REG_MCR:
+        val = s->uart_mcr;
+        break;
+
+    case UART_REG_LSR:
+        val = 0x60 | s->uart_rx_pending;        /* tx always ready/empty */
+        break;
+
+    case UART_REG_MSR:
+        val = 0xB0;                             /* report CTS, DSR, DCD as asserted */
+        break;
+
+    case UART_REG_SCR:
+        val = s->uart_scr;
+        break;
+
+    default:
+        val = 0;
+        break;
+
+    }
+    return val;
+}
+
+static void console_write_char(RISCVMachine *s, uint8_t c)
```

Comment:
This might benefit from inlining:

```suggestion
static inline void console_write_char(RISCVMachine *s, uint8_t c)
```